### PR TITLE
Aave Funds: Acquire B-80-BAL-20-WETH BPT

### PR DESCRIPTION
**Pull request by efecarranza**
# Changelog

Add payload to handle approvals and transfers of tokens.
Add COWTrader contract to handle trading and cancellation of trades for swapping tokens.
Add tests for both contracts.

Please see the following HackMD document for QA and research regarding this trade: https://hackmd.io/hNA14m6zSi68Y2FQQloJWA

### CODE REVIEW

File: src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+   b. Withdraw aEthBAL
+   c. Transfer all BAL now held in Collector to COWTrader.sol
+   d. Transfer specified wETH quantity to COWTrader.sol
+2. The COWTrader contract has one function to execute two trades using Milkman, utilizing its balance of wETH and BAL respectively. Both trades take the held token for 80-BAL-20-WETH BPT.
+3. Milkman creates a new instance per trade, which will then hold the funds until the trades are settled via CoW Swap. Under the hood, Milkman approves CoW Swap's Settlement contract so it can transfer funds once there is a match.
+4. Once there is a match, the dynamic price checker will protect Aave's funds against a bad quote. If it passes our specified parameters, then the trade will go through and the Collector contract will receive 80-BAL-20-WETH.
+
+If everything goes well and the number of tokens to be received matches our parameters (slippage, price against our oracle) then the Collector contract will receive the BPT.
+
+The trade can remain open in perpetuity until it is canceled (by Aave) or settled.
```
Comment (eboadom) [inline src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md:71]: The cancelling is not by Aave, no?

File: src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+# Implementation
+
+A list of relevant links for this proposal:
+
+- [Governance Forum Discussion](https://governance.aave.com/t/arfc-deploy-ethereum-collector-contract/12205)
+- [Test Cases](XXX)
```
Comment (eboadom) [inline src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md:78]: Missing

File: src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+- [Payload Implementation](XXX)
```
Comment (eboadom) [inline src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md:79]: Missing

File: src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
+---
+title: Treasury Management - Acquire B-80BAL-20WETH
+discussions: https://governance.aave.com/t/deploy-bal-abal-from-the-collector-contract/9747
+shortDescription: Aave DAO is to swap BAL and wETH to B-80BAL-20WETH via CoW Swap using Milkman to prevent MEV attacks and bad price slippage whilst mitigating reentrancy risk.
```
Comment (eboadom) [inline src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md:4]: What this refers to? `whilst mitigating reentrancy risk`

Comment (efecarranza) [reply src/AaveV2CollectorSwap80BAL20WETH_20230403/SWAP-ASSETS-FOR-BPT.md]: Was just referring to the fact that we're checking to see if a trade already exists as well as functions only callable by governance/allowed caller. Probably unnecessary just wanted to highlight for the forum what we had taken into account.

**Review by sakulstra — APPROVED** [review]

**Pull request merged**